### ISSUE TITLE
Autofill always asks for master password, unless "session timeout" is set to "never"

### ISSUE THREAD

**Issue by michaelwimmer** [bitwarden/ios#1167]
### Steps To Reproduce

1. Set up autofill as requested in the bitwarden app
2. Set the "Session timeout" to 15 minutes
3. Make sure "Unlock with Face ID" and "Unlock with PIN code" are deactivated
4. Go to a website with a login
5. Try to autofill (tapping e.g. the "Passwords" button on the onscreen keyboard): the Bitwarden autofill window opens.
6. Enter the master password
7. Go to another website login
8. Tap on 'Passwords' before the 15 minutes have passed
9. Bitwarden autofill window opens again, asking again for the master password

### Expected Result

Bitwarden should not ask for the master password again, if the second time it is opened falls into the window set by "Session timeout".

### Actual Result

Bitwarden always asks for the master password. 

The only way I could avoid this is by setting in  step 2 "Session timeout" to "never". Then autofill never asks for the master password.

However, this is obviously not an ideal setting.

### Additional Context

This behavior of the bitwarden autofill is especially annoying for logins that span two pages, such as gmail. Then you have to enter the master password twice for one login.

One can find similar reports e.g. on reddit (see below) since quite a while. Typically, the advice is to use biometrics. Indeed, if I enable biometrics, bitwarden will scan my face every time in autofill. However, I would prefer to avoid using Face-ID for various reasons. Also, given that "never" actually works as expected, any other session timeout should work, too.

Examples of similar issues reported online:
- https://www.reddit.com/r/Bitwarden/comments/zlp9an/constantly_having_to_reenter_master_password_ios/
-  https://www.reddit.com/r/Bitwarden/comments/10g0m8k/ios_app_bitwarden_prompts_for_master_password/

### Build Version

Version 2024.11.0 (1680)

### What server are you connecting to?

EU

### Environment Details

- Device: IPad Pro (11-inch, second generation)
- IPadOS Version: 18.1
- 

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

**Comment by francoposa** [comment]
+1 I am also experiencing this. The autofill is really the only way I ever interact with Bitwarden on the phone, so this issue is not an afterthought, it is really frustrating.

**Comment by mudtar** [comment]
I am also experiencing the same issue.

**Comment by kholisrag** [comment]
have same issue here, I have iOS with disabled FaceID, the iOS autofill always ask for master password instead off PIN

**Comment by tanguylegazon** [comment]
I have the same issue on iOS:
- iPhone XR
- iOS 18.2 (22C152)
- Bitwarden 2024.12.0 (1740)

**Comment by mel-m** [comment]
I have the same issue and hope this gets resolved soon. I have Face ID enabled but even needing to do a face scan multiple times is too unusable.


Version: 2024.12.0 (1740)

-------- Device --------

Model: iPhone14,7
OS: iOS 18.1.1

------- CI Info --------

Branch: release/2024.12-rc1
CI Run Attempt: 1
CI Run ID: 12126636763
CI Run Number: 433
Commit hash: c621b02e207971eee21b1421f210f820e9ebbbda
Compiler Flags: 
Repository: https://github.com/bitwarden/ios

**Comment by mtalhakrc** [comment]
+1. Autofill is asking for confirmation on every attempt without respecting to the session timeout

**Comment by Gerardv514** [comment]
I can reproduce as well very annoying which leads me to set time out to never. 

I Wish there was another option for timeout lock "when device is restarted"

**Comment by thekevincurtin** [comment]
Same issue, have been encountering this for over a year

**Comment by PCSerra** [comment]
Same here, autofill with faceID disabled on iOS, session timeout AND pin settings are ignored and master password needed every time.

**Comment by matteoiurato** [comment]
+1

**Comment by mtalhakrc** [comment]
so do we have any updates?

**Comment by killakalle** [comment]
Experienced the same on my father in law's iPhone

**Comment by micouy** [comment]
+1, it makes BitWarden unusable as a replacement for the default passwords app in iOS. I enabled PIN (but not biometrics) and set the timeout to when the app is restarted. I can open the app and see all passwords fine but it still prompts me for my master password every time I use autofill

**Comment by Gerardv514** [comment]
> +1, it makes BitWarden unusable as a replacement for the default passwords app in iOS. I enabled PIN (but not biometrics) and set the timeout to when the app is restarted. I can open the app and see all passwords fine but it still prompts me for my master password every time I use autofill 

As per my conversation [here](https://github.com/bitwarden/ios/issues/1634#issuecomment-2956753634) with @fedemkr i believe this is expected behavior. Despite having the vault timeout to 4 hrs or app restart when autofilling it will always call for master password (or pin or biometric) for the autofill.

**Comment by micouy** [comment]
> If you have iOS 18+, which from what I see you have, you can either:
> - Use Autofill TOTP directly if the site has configured that correctly.
> - **Use the more generic Long press text Autofill (Long press on the field you want to autofill to open the context menu, then go to "Autofill" and then "Passwords") which will open the Bitwarden extension app for you to choose the item and what you want to autofill, in this case the TOTP.**

Even when I do this, BW still prompts me for the master password, even though the BW app is unlocked and running in the background. It's easier (but still poor UX) to just go to BW app directly instead of using the AutoFill feature.

**Comment by Gerardv514** [comment]
> > If you have iOS 18+, which from what I see you have, you can either:
> > 
> > * Use Autofill TOTP directly if the site has configured that correctly.
> > * **Use the more generic Long press text Autofill (Long press on the field you want to autofill to open the context menu, then go to "Autofill" and then "Passwords") which will open the Bitwarden extension app for you to choose the item and what you want to autofill, in this case the TOTP.**
> 
> Even when I do this, BW still prompts me for the master password, even though the BW app is unlocked and running in the background. It's easier (but still poor UX) to just go to BW app directly instead of using the AutoFill feature.

Keep reading the next few comments:

I [asked ](https://github.com/bitwarden/ios/issues/1634#issuecomment-2957141107)if this was expected, and as per fedemkr it [is](https://github.com/bitwarden/ios/issues/1634#issuecomment-2959091081).

**Comment by micouy** [comment]
Your comment is about Face ID which I have disabled. If auth is needed every time, I enabled PIN but BW is still prompting for the master password.

**Comment by Gerardv514** [comment]
Face ID and pin are the same, they are both Auth. And even so the behavior you mentioned happens with both. So if @fedemkr is saying it is expected behavior to have to Auth when autofilling , whether it's Face ID or pin; it's expected that you have to do that.

**Comment by roundish594** [comment]
> Face ID and pin are the same, they are both Auth. And even so the behavior you mentioned happens with both. So if [@fedemkr](https://github.com/fedemkr) is saying it is expected behavior to have to Auth when autofilling , whether it's Face ID or pin; it's expected that you have to do that.

Agree, but only outside of the selected session timeout. Within the selected session timeout it should not ask for auth at all. What should be the sense behind session timeout if auth is always required?

**Comment by Gerardv514** [comment]
I agree too, I don't like that it request to authenticate when autofilling, especially when still within that session window. This is the reason why I leave my timeout set to never. I don't know if it has something to do with iOS and this needs to be functioning in this way. But as I had asked in the other thread for something else it seemed like from the response this is expected behavior. It's a bit of a nuisance cause it slows the flow down and I already have a heavy Auth to unlock the phone.

**Comment by roundish594** [comment]
I think it is not something from iOS. I have switched from KeePassium where it is possible to unlock within the session window without auth. As this is also Open Source, maybe someone can have a look how they have solved it (in case it is a bug and not a feature).

**Comment by micouy** [comment]
> As per my conversation https://github.com/bitwarden/ios/issues/1634#issuecomment-2956753634 with @fedemkr i believe this is expected behavior. Despite having the vault timeout to 4 hrs or app restart when autofilling it will always call for master password **(or pin or biometric)** for the autofill.

> **Face ID and pin are the same**, they are both Auth. And even so the behavior you mentioned happens with both. So if @fedemkr is saying it is expected behavior to have to Auth when autofilling , whether it's Face ID or pin; it's expected that you have to do that.

I'm fine with being asked for PIN and I enabled it, but it's still asking me for the master password. Seems like it's not the expected behavior.

**Comment by DigiDr** [comment]
This should not be the expected behaviour.... 

- What we expect: with FaceID enabled, Master Password is not prompted for unless we've hit the session limit or logged out of the account. 

- Current behaviour: with FaceID enabled, Master Password is asked for every single time we call on bw for a password, unless we set timeout to never (this is unsafe). 

@Gerardv514 asked @fedemkr "if I set a 4 hour time out to lock, and I turn on Face ID to unlock, is it expected that bw will activate Face ID anytime I do an autofill? (Using the autofill above iOS keyboard)." 

Yes, this is what we expect to happen. FaceID should always be prompted for if enabled. This is not the correct question for this thread, however. What we do not expect to happen is that it ALSO asks for the Master Password.

**Comment by mtalhakrc** [comment]
It's really interesting that they haven't been able to resolve this issue for such a long time, let alone provide any explanation. It makes me wonder if there are no developers available to maintain the project. Just think about how much time has passed since this issue was first opened...

I wish I knew Swift so I could try to resolve this with vibe code stuff.

**Comment by kholisrag** [comment]
got this reply from bitwarden support, no ETA 

![Image](https://github.com/user-attachments/assets/56859d25-0671-4c3c-b897-53dda859b320)

**Comment by DigiDr** [comment]
Thanks @kholisrag - the explanation seems to suggest it's unfixable without some kind of workaround - if we don't want timeout set to forever. Interesting!

**Comment by Gerardv514** [comment]
> got this reply from bitwarden support, no ETA 
> 
> ![Image](https://github.com/user-attachments/assets/56859d25-0671-4c3c-b897-53dda859b320)

Odd, when autofilling Bitwarden always defaults to a locked state due to memory? 


This doesn't happen if you set vault timeout to never then you are never prompted for Auth when autofilling, whether it's biometric or master password. So I guess somehow iOS knows to keep the bw autofill in an unlocked state for good.

**Comment by Gerardv514** [comment]
> I'm fine with being asked for PIN and I enabled it, but it's still asking me for the master password. Seems like it's not the expected behavior.

Agreed. Being asked for the master password doesn't happen for me (when Face ID or pin are enabled). If you are being prompted for a master password when having either pin or Face ID enabled that is not expected behavior.

Separately, i also agree you shouldn't be prompted for Face ID, pin, or master password when autofilling if you're within a defined session time. The screenshot above from Bitwarden explains why, and is due to iOS memory and the session always defaulting to a locked state. 

I'd question why doesn't it happen when the timeout of the vault is set to never, it respects that and stays unlocked.

**Comment by Gerardv514** [comment]
@kholisrag can you reply back to that Bitwarden support email, asking what I had mentioned. 

If Bitwarden by default goes into a locked state, why doesn't that happen when the vault timeout is set to never.

**Comment by kholisrag** [comment]
let me replied it, and ask them.

**Comment by Gerardv514** [comment]
@kholisrag any response from them?

**Comment by kholisrag** [comment]
They only response like this,

![Image](https://github.com/user-attachments/assets/28394d7a-f5be-4db4-a83a-10e34d05dc59)

**Comment by Gerardv514** [comment]
🤦‍♂️ seems like the responses are on opposite sides of each other. First one indicates nothing wrong and function as normal, second is there's possibly something not working right. 

Either way hoping there will be a fix, for all scenarios (constantly prompted for master password, and pin/face authentication not respecting the session timeout when set to anything other than never).

**Comment by farkastime** [comment]
This really is a critical bug folks, as there's currently no way to specify a reasonable timeout for the entry of a master password, which ought to be a PIA to enter if chosen properly. Sadly, I'll have to revert to my old password manager until this is fixed.

**Comment by t-moe** [comment]
This bug is a real blocker for everyday use on iOS , especially for non-technical users.

Having to enter the master password every single time for autofill (regardless of session timeout settings) makes Bitwarden feel unusable for some, and in my case it’s already discouraging family members from adopting a password manager at all.

I understand iOS autofill has technical constraints, but this issue has been open for a long time without visible progress. Could the team please raise its priority and share a plan or timeline? This one directly impacts adoption and retention.

**Comment by Gerardv514** [comment]
Previously I had left the timeout on the vault set to never, but recently a change was made that now prevents the pasting of the Totp code after autofill. 

So now have to use a timeout despite a strong password on the iOS device. There are certain sites that utilize a separate page for user name , password , and Totp code...meaning you need to enter your long and complex master password 3 x's... this workflow essentially will cause users to weaken their security; making their master password much simpler as one example. 

So far I've used time out when app restarts and have been using faceid to get around it..takes a bit longer but much quicker than typing the master password, and the pasting of Totp codes work only when you have a timeout. Also the Face ID authentication will mimic the master password requirement. Requiring a Face ID for user name auto fill, pass work, and Totp.

**Comment by yinso** [comment]
+1 on this issue. the only setting that doesn't prompt for master password for me is "Never". This renders Bitwarden either extremely insecure or unusable on iPhone with long master passwords.

**Comment by Gerardv514** [comment]
Hello @KatherineInCode can anything be done about this?

**Comment by fedemkr** [comment]
Hello there, we're currently discussing on how to improve the UX on this so users don't have to enter the master password every time when not using Pin nor Biometrics. 
However, it's not that straightforward in order to maintain the high security of the app and in this case, the extensions (the Autofill flows run in an iOS extension).
I'll post more updates as soon as I have more news on this.
Thank you all for the feedback.

**Comment by Gerardv514** [comment]
@fedemkr i would think if a user is not using biometrics they should be prompted for a master password to keep the app secure. 

The issue is when biometrics is enabled, it asks for the master password prior to the timeout session expiring. Instead it should utilize the biometrics for the authentication, until after the timeout is reached then probably should be prompted again for master password. Any subsequent attempts to auto fill, while within the timeout period, should authenticate by biometrics again.

**Comment by tanguylegazon** [comment]
> [@fedemkr](https://github.com/fedemkr) i would think if a user is not using biometrics they should be prompted for a master password to keep the app secure.

I’m not really sure I agree with that. I don’t use Face ID to unlock my Bitwarden vault, but I still configured a 15-minute session timeout in the app. I don’t see how applying the same rule to autofill would make things any less secure. That said, I’m not commenting on the feasibility — I know Apple makes things pretty restrictive when it comes to iOS extensions.

**Comment by Gerardv514** [comment]
@tanguylegazon so after the 15 min time out how do you unlock?

It's either biometric, pin , or master password. 

I probably should have included pin as well and not solely biometric. 

Point is if your within the timeout you shouldn't keep getting prompted for some sort of authentication, which is what is happening except it is always prompting for master password regardless if you use biometric or pin to unlock.

**Comment by KatherineInCode** [comment]
The primary issue is maintaining the timeout between the main app and the extension, because they don't share the same memory space, and the way timeout is implemented is (reductively speaking) around how long we let certain things exist in memory. As my colleague said, it's something we've been discussing, and working towards a secure solution.

**Comment by fedemkr** [comment]
@Gerardv514 if you're seeing an issue with Biometrics enabled then please open a new issue as this one is about **not** having Biometrics nor PIN enabled on Bitwarden. That's how the report is stated in **Steps to Reproduce**, check point (3):
 
> Make sure "Unlock with Face ID" and "Unlock with PIN code" are deactivated

I'll check about the Biometrics issue you're mentioning though but if you could add a new issue with the information I'd appreciate it as the scope of the problem is not the same.

**Comment by Gerardv514** [comment]
Sorry I'm mixing up threads. I retested with a timeout and it's using the biometrics. Previously it was always prompting for master password.

**Comment by mattf9** [comment]
I'm also encountering this issue on iOS. Works fine on Mac OS, linux, etc. This is a showstopper for me - I cannot use Bitwarden with this broken functionality. I do NOT want to use pin or biometrics. I want to Set the "Session timeout" to 15 minutes and have it function as expected.

**Comment by cpilson** [comment]
+1. This is a major bug and a deviation from anticipated behavior. 

It sounds like the vault is properly not being locked immediately on auto-fill for biometric users, but locks immediately after auto-fill use for password users.

**Comment by bbbb4tryon** [comment]
I’ve been lurking here awhile. This is not a rant, it is designed to illustrate all the trouble shooting and self-help I’ve performed from both the comments here, the mod’s instructions and the Support Desk’s responses. The comment is intertwined with identified gaps encountered While trouble shooting.

I'm also encountering this issue on everything: Mac with Sequoia; iPhone with 26.1. Safari mobile, Chrome desktop, Brave mobile, Brave desktop. This happened before too, needed help troubleshooting. Currently (Dec 2025), I followed the troubleshooting steps and they Have Not fixed the issue: 
iPhone - safari, Brave
- 1. log out Bitwarden extension, remove it from extensions; 
- 2. quit browser, 
- 3. restart browser; 
- 4. download extension; 
- 5. log back in with master password; 
- 6. set to 4 min: go to website that requires username + password; 
- 7. must enter master password 

the Bitwarden app: 
Log out
Delete app from device 
Clear cache
Restart phone
Download app
Log into app with master password 
Set to 4 min: go to website that requires username + password
Must enter master password 

Repeat every step with same result on Mac; desktop Chrome; Chrome with Bitwarden extension; Brave desktop; Brave with desktop extension 

I don’t know enough about security specifics except that a master password is basically enough. That’s what Bitwarden told us and sold us on: and quibbling about specifics is dumb. A master password and vault are the reason. Most people neither know nor care about How Bitwarden keeps us safe, they only care that it does and that there is NO friction in use.

Why back when, Norton told is Security is compromised the more openings there are. That is: the more options (not more layers), the more cracks bad things can seep through. An “additional” pin or biometrics or passkey or ANY additional “help” is like writing down the password in the PASSWORDS notebook in the bottom drawer.

I want to set the "Session timeout" to HOURS from now - like the whole workday because that’s why the extension pops up… it’s a secure because people can’t Also logon to the phone or Mac Unless they have my face and/or my complicated password that I don’t keep written down on my phone or post it note. I imagine most Bitwarden users want the program to hold and fill all their passwords with a click. In the mobile case, One click to copy the long master password and then having to back up from the login and move back into it to click the login, then open and paste the master to then fill the login and password are too many steps and even writing all the steps is long and infuriating. 

Please please please: frictionless, long-time, no extra biometrics, no more than one or two copy paste. This hopefully doesn’t become Apple where, now, function is ignored and pretty is promoted.

**Comment by KatherineInCode** [comment]
I have been starting to work on the groundwork for things that should eventually allow us to fix this more permanently. The current PR that I am currently working on (and still building out, as of the time of this posting) is https://github.com/bitwarden/ios/pull/2239.

**Comment by Gerardv514** [comment]
> I have been starting to work on the groundwork for things that should eventually allow us to fix this more permanently. The current PR that I am currently working on (and still building out, as of the time of this posting) is https://github.com/bitwarden/ios/pull/2239.

Would this also allow for the ability to copy/paste a totp code after an autofill? Currently totp codes won't copy if timeout is never too.

**Comment by KatherineInCode** [comment]
> Would this also allow for the ability to copy/paste a totp code after an autofill?

I am currently just working on how sessions are handled between the PM app and the AutoFill extension. I think that is a separate issue, which while I do eventually want to get fixed, is not my current focus.

**Comment by Gerardv514** [comment]
Awesome thanks for everything!

**Comment by vinovod** [comment]
Almost 2 years to fix real showstopper. Unbelievable.

**Comment by thekevinc** [comment]
can someone just ask claude to fix it and raise a PR

**Comment by pamperer562580892423** [comment]
@thekevinc:

> can someone just ask claude to fix it and raise a PR

I think this is currently being worked on - see [this comment](https://github.com/bitwarden/ios/issues/1167#issuecomment-3725930882) from @KatherineInCode.

**Comment by Gerardv514** [comment]
So this has been pushed out in a release 2026.2.0, but is it just me I am not seeing a difference and the issue is still happening? Maybe due to more additional work that needs to be done in other future PRs?

**Comment by KatherineInCode** [comment]
Yes, as I said, #2239 was just laying some groundwork to eventually fix this. #2323 likewise is laying additional groundwork. We *are* moving towards a solution for this and some related issues, we're just making sure each step of the way that we keep things secure. It's a tricky problem space, making two processes act like they're one when it comes to "session" without creating a vulnerability.

**Comment by Gerardv514** [comment]
Beautiful that makes sense and is worth the wait!! Thank you again for the hard work.

**Comment by pamperer562580892423** [comment]
@fedemkr wrote [some time ago](https://github.com/bitwarden/ios/issues/1167#issuecomment-3309757356):

> [@Gerardv514](https://github.com/Gerardv514) if you're seeing an issue with Biometrics enabled then please open a new issue as this one is about **not** having Biometrics nor PIN enabled on Bitwarden.

@michaelwimmer could you add something like "(biometric and PIN unlock both disabled)" to the title of this issue to highlight this special - and maybe uncommon - condition?

(I called it "uncommon" because I would guess most people probably do have at least one of those two unlock options enabled)

**Comment by michaelwimmer** [comment]
I amended the title. I don't think it is that uncommon though ;)

**Comment by p1100i** [comment]
Hello @michaelwimmer and @pamperer562580892423 !

I can reproduce this with only PIN unlock enabled. So I think the title is misleading, as having PIN unlock enabled, you will be still always reprompted for login with master password.

Enabling TouchID kind of fixes the behavior, but you can still see the session is not "kept" alive, but bitwarden just quickly uses fingerprint based auth to recreate a session.

Version: 2026.5.0 (3226) on iOS 26.4.2 using Safari

EDIT: proper wording.

**Comment by pamperer562580892423** [comment]
@p1100i ... then open a new issue about that as @fedemkr (a BW developer) suggested here: https://github.com/bitwarden/ios/issues/1167#issuecomment-3309757356

I would add the info there that it shouldn't be considered a duplicate of this issue, as per @fedemkr's explicit request to open a separate issue.

PS: The new title now is less misleading, IMHO, as it just represents the original issue that BW investigates currently.

**Comment by KatherineInCode** [comment]
> PIN unlock enabled

I think it is worth noting that currently, there are two different forms this can take. When you turn "Unlock with PIN code" on, you are asked "Do you want to require unlocking with [biometrics] or your master password when the application is restarted?", and both of these show as "on", but potentially produce different results vis-à-vis requesting the PIN code in the AutoFill extension. When noting in a new ticket that PIN unlock is enabled, it would be very helpful to know whether it's in the "require biometrics/MP on application restart" state or not.

That said, we *are* having discussions internally about this (along with soliciting some community feedback in places such as [Reddit](https://www.reddit.com/r/Bitwarden/comments/1tq3p60/lets_talk_about_on_app_restart_vault_timeout_on/)), and driving to solutions that hopefully solve this whole class of issues, while being simpler to reason about (both as a user and in code) and continuing to maintain the security of the app.

**Comment by michael-herbst** [comment]
I spent some time systematically testing this behaviour on my iPad and wanted to share the results, as they may help narrow down the issue.
I can confirm that this issue is still reproducible with the current release.

Environment

- Device: Apple iPad (10th generation)
- Bitwarden Version: 2026.7.0 (3402)
- Latest iPadOS (August 2026)

Configuration

- Vault Timeout Action: Lock
- Unlock Method: Touch ID (same behaviour reproduced using PIN)

Results

| Vault Timeout | Result |
|---------------|--------|
| 1 minute | Touch ID required for every AutoFill |
| 15 minutes | Touch ID required for every AutoFill |
| 1 hour | Touch ID required for every AutoFill |
| Never | AutoFill behaves exactly as expected without repeated authentication |

Additional testing

- Disabled Bitwarden as AutoFill provider
- Restarted the iPad
- Re-enabled Bitwarden AutoFill

No change.

The Bitwarden vault itself remains unlocked as expected.

The repeated authentication only affects the iPadOS AutoFill extension.

The behaviour appears to depend solely on whether the Vault Timeout is finite or set to Never.

This strongly suggests that the AutoFill extension is not honoring the configured Vault Timeout.

**Comment by Gerardv514** [comment]
It's been mentioned before, and your observations are correct. This is due to the iOS Bitwarden app and the Bitwarden extension do not share the same memory space (or something near that realm). This is being worked on by @fedemkr & @KatherineInCode and their team.
